### ISSUE TITLE
Bug when using Anonymized DNSCrypt & ODoH at the same time

### ISSUE THREAD

**Issue by Zastoff** [DNSCrypt/dnscrypt-proxy#1992]
THE TRACKER IS DEDICATED TO KEEPING TRACK OF *BUGS*,
preferably after they have been already discussed and confirmed to be reproducible.

FOR ASSISTANCE, PLEASE CLOSE THIS FORM AND USE THE DISCUSSIONS SECTION INSTEAD:
https://github.com/DNSCrypt/dnscrypt-proxy/discussions/categories/q-a

~~~

Reported bugs must reproducible in the context described in the "Context" section.

Installation and configuration issues are not bugs, but individual assistance request.

Context: the LATEST version of `dnscrypt-proxy` (precompiled binaries downloaded from this repository) is correctly installed and configured on your system, but something doesn't seem to produce the expected result.

If the bug is not trivial to reproduce on any platform, please include ALL the steps required to reliably duplicate it, on a vanilla, generic install of macOS, Windows, OpenBSD or Ubuntu Linux system, in their most current version.

If you don't have any clear understanding of the issue or can't enumerate the steps to reproduce it, open a discussion instead:
https://github.com/DNSCrypt/dnscrypt-proxy/discussions


## Output of the following commands:

./dnscrypt-proxy -version
izzt@RT-AX88U-6C58:/jffs/dnscrypt# ./dnscrypt-proxy -version
2.1.1
./dnscrypt-proxy -check
izzt@RT-AX88U-6C58:/jffs/dnscrypt# ./dnscrypt-proxy -check
[2022-01-15 10:15:27] [NOTICE] dnscrypt-proxy 2.1.1
[2022-01-15 10:15:27] [NOTICE] Dropping privileges
[2022-01-15 10:15:27] [NOTICE] Source [public-resolvers] loaded
[2022-01-15 10:15:27] [NOTICE] Source [relays] loaded
[2022-01-15 10:15:27] [NOTICE] Source [odoh-servers] loaded
[2022-01-15 10:15:27] [NOTICE] Source [odoh-relays] loaded
[2022-01-15 10:15:27] [NOTICE] Anonymized DNS: routing [odoh-cloudflare] via [odohrelay-koki-se odohrelay-koki-ams]
[2022-01-15 10:15:27] [NOTICE] Anonymized DNS: routing [dnscrypt.eu-nl] via [anon-cs-swe anon-sth-se anon-acsacsar-ams-ipv4 anon-cs-dk]
[2022-01-15 10:15:27] [NOTICE] Anonymized DNS: routing [serbica] via [anon-ams-nl anon-sth-se anon-cs-swe]
[2022-01-15 10:15:27] [NOTICE] Anonymized DNS: routing [dct-ru1] via [anon-sth-se anon-cs-finland anon-meganerd]
[2022-01-15 10:15:27] [NOTICE] Anonymized DNS: routing [meganerd] via [anon-sth-se anon-cs-swe anon-opennic-R4SAS]
[2022-01-15 10:15:27] [NOTICE] Anonymized DNS: routing [opennic-R4SAS] via [anon-cs-norway anon-sth-se anon-meganerd]
[2022-01-15 10:15:27] [NOTICE] Anonymized DNS: routing [sth-dnscrypt-se] via [anon-cs-swe anon-cs-norway anon-meganerd]
[2022-01-15 10:15:27] [NOTICE] Anonymized DNS: routing [odoh-koki-ams] via [odohrelay-koki-se odohrelay-crypto-sx]
[2022-01-15 10:15:27] [NOTICE] Anonymized DNS: routing [odoh-koki-se] via [odohrelay-crypto-sx odohrelay-koki-ams]
[2022-01-15 10:15:27] [NOTICE] Configuration successfully checked
./dnscrypt-proxy -resolve example.com
izzt@RT-AX88U-6C58:/jffs/dnscrypt# ./dnscrypt-proxy -resolve example.com
Resolving [example.com] using 127.0.1.1 port 53

Resolver      : 74.125.74.12

Canonical name: example.com.

IPv4 addresses: 93.184.216.34
IPv6 addresses: 2606:2800:220:1:248:1893:25c8:1946

Name servers  : a.iana-servers.net., b.iana-servers.net.
DNSSEC signed : yes
Mail servers  : 1 mail servers found

HTTPS alias   : -
HTTPS info    : -

Host info     : -
TXT records   : yxvy9m4blrswgrsz8ndjh467n2y7mgl2, v=spf1 -all

- [ ] Initially raised as discussion #...
From Syslog:
Jan 15 06:59:54 RT-AX88U-6C58 dnscrypt-proxy[15940]: Anonymizing queries for [odoh-koki-ams] via [odohrelay-crypto-sx]
Jan 15 06:59:58 RT-AX88U-6C58 dnscrypt-proxy[15940]: Anonymizing queries for [odoh-cloudflare] via [anon-sth-se]<--?
Jan 15 07:00:00 RT-AX88U-6C58 dnscrypt-proxy[15940]: Anonymizing queries for [odoh-koki-se] via [anon-ams-nl]<--?
and i get these:
Jan 15 09:58:59 RT-AX88U-6C58 dnscrypt-proxy[15940]: Failed to receive successful response from [odoh-koki-se]
From toml:
{ server_name='odoh-koki-se', via=['odohrelay-crypto-sx', 'odohrelay-koki-ams'] }
{ server_name='odoh-cloudflare', via=['odohrelay-koki-se', 'odohrelay-koki-ams'] },
and so on: I do not have any dnscrypt relays set for ODoH servers but when the proxy is starting it tries to use them it seems


## *What* is affected by this bug?
ODoH and Anonymized DNSCrypt configured at the same time


## *When* does this occur?
When the proxy starts it checks relays but it seem too try dnscrypt relays for ODoH sometimes even if it not set for it in toml


## *Where* does it happen?
On start and restart


## *How* do we replicate the issue?
<!-- Please list all the steps required to reliably replicate it, starting from a newly installed operating system -->



## Expected behavior (i.e. solution)



## Other Comments

**Comment by jedisct1** [comment]
Thanks for reporting this!

What does the entire `routes` section look like? Just wondering where the Anonymized DNSCrypt relays are defined in your configuration.

**Comment by Zastoff** [comment]
Looks like this:
`routes = [ { server_name='dct-ru1', via=['anon-sth-se', 'anon-cs-finland', 'anon-meganerd'] }, { server_name='dnscrypt.eu-nl', via=['anon-cs-swe', 'anon-sth-se', 'anon-acsacsar-ams-ipv4', 'anon-cs-dk'] }, { server_name='meganerd', via=['anon-sth-se', 'anon-cs-swe', 'anon-opennic-R4SAS'] }, { server_name='opennic-R4SAS', via=['anon-cs-norway', 'anon-sth-se', 'anon-meganerd'] }, { server_name='serbica', via=['anon-ams-nl', 'anon-sth-se', 'anon-cs-swe'] }, { server_name='sth-dnscrypt-se', via=['anon-cs-swe', 'anon-cs-norway', 'anon-meganerd'] }, { server_name='odoh-cloudflare', via=['odohrelay-koki-se', 'odohrelay-koki-ams'] }, { server_name='odoh-koki-ams', via=['odohrelay-koki-se', 'odohrelay-crypto-sx'] }, { server_name='odoh-koki-se', via=['odohrelay-crypto-sx', 'odohrelay-koki-ams'] } ]`

skip_incompatible = true

direct_cert_fallback = true

**Comment by jedisct1** [comment]
Hi!

I was not able to reproduce this issue.

Can you share a minimal configuration file that reliably reproduces it?

What version are you running?

**Comment by jedisct1** [comment]
Are you using `[static]` entries by chance?

**Comment by Zastoff** [comment]
No static servers or relays with this setup.
Think the installer i use installs this for my router: dnscrypt-proxy-linux_arm64-2.1.1.tar.gz
and this is the installer and what i use to configure it with: https://github.com/thuantran/dnscrypt-asuswrt-installer

**Comment by Zastoff** [comment]
See if this works
https://file.io/rahl00oPZOVL
it`s the dnscrypt-proxy.toml i use

**Comment by jedisct1** [comment]
"The file you requested has been deleted"

**Comment by Zastoff** [comment]
Try this instead
https://1drv.ms/u/s!Am64iqBZS2NSah1RsE17s7Dy2Uc?e=9XWY3v

**Comment by jedisct1** [comment]
`anon-ams-nl` and `odohrelay-koki-ams` are exactly the same server (it handles both protocols).

**Comment by jedisct1** [comment]
So, that was mainly a cosmetic issue. It was still using a relay that can handle the right protocol.

It should be fixed in  06733f5 (and so, will be in the next version).

Thank you for reporting this!

**Comment by Zastoff** [comment]
> So, that was mainly a cosmetic issue. It was still using a relay that can handle the right protocol.
> 
> It should be fixed in [06733f5](https://github.com/DNSCrypt/dnscrypt-proxy/commit/06733f57edce284ae287ccd77feeb45a99571023) (and so, will be in the next version).
> 
> Thank you for reporting this!

Thank you for solving it !
